Test task

ErmakovDenis opened the pull request.

ErmakovDenis committed:
Add ci.yml Dockerfile nginx.conf

ErmakovDenis committed:
Fix error

ErmakovDenis committed:
Docker always build and push

ErmakovDenis committed:
try fix error

ErmakovDenis committed:
try fix error 2

ErmakovDenis committed:
try fix error 3

ErmakovDenis committed:
Add docker compose

ErmakovDenis committed:
Add docker compose

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+name: Pipeline
+on: [push]
```
pls restrict the trigger to
- open PR against master branch
- pushes against a specific branch. So let's say you develop feature/X - you may not want to trigger the build (as it's a work in progress) until it's officially merged into develop branch.

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+        run: cd angular/ && npm install
+      - name: Run tests
+        run: cd angular/ && npm run coverage
+
+
```
remove spaces.

dmitry-mightydevops replied:
if you want spaces between steps - this is fine, but then be consistent: somewhere you use one or two space lines.

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+      - name: Set up Node.js
+        uses: actions/setup-node@v2
+        with:
+          node-version: '18'
```
don't hardcode the version, developers typically modify package.json to include the version requirements
https://github.com/ErmakovDenis/angular-react-starter/blob/10061ddfa8cc63fd063a543c43bea7b97681ad0d/angular/package.json#L3

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+      - name: Install dependencies
+        run: cd angular/ && npm install
+      - name: Run tests
+        run: cd angular/ && npm run coverage
```
let's setup setup working directory for your job. instead of using `cd` cli in steps.

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+        run: |
+          echo "STOP_TIME=$(date +%s)" >> $GITHUB_ENV
+
+      - name: Check environment variables
+        if: always()
+        run: |
+          START_TIME=$(echo ${{ env.START_TIME }})
+          STOP_TIME=$(echo ${{ env.STOP_TIME }})
+          EXECUTION_TIME=$((STOP_TIME - START_TIME))
+          HOURS=$((EXECUTION_TIME / 3600))
+          MINUTES=$((EXECUTION_TIME % 3600 / 60))
+          SECONDS=$((EXECUTION_TIME % 60))
+          echo "EXECUTION_TIME_HOURS=$HOURS" >> $GITHUB_ENV
+          echo "EXECUTION_TIME_MINUTES=$MINUTES" >> $GITHUB_ENV
+          echo "EXECUTION_TIME_SECONDS=$SECONDS" >> $GITHUB_ENV
+          echo "Execution time: $HOURS hours $MINUTES minutes $SECONDS seconds"
```
lets refactor timing calculation this into using `gh cli`, https://cli.github.com/
your code will not need start/stop time recorded vars and your pipeline would become nicer.

kseniyashaydurova replied:
Don't see that it is fixed

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+        if: success()
+        uses: dawidd6/action-send-mail@v3
+        with:
+          server_address: smtp.gmail.com
```
should not be hardcoded, should be passed from **inputs**
see in docs how to use
on.workflow_dispatch.inputs and introduce your variables on top that you use inside your pipeline

kseniyashaydurova replied:
Don't see that it was fixed

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+          body: |
+            Pipeline success! Finally!
+            Final execution time: ${{ env.EXECUTION_TIME_HOURS }} hours ${{ env.EXECUTION_TIME_MINUTES }} minutes ${{ env.EXECUTION_TIME_SECONDS }} seconds
+          to: clutcherdenis@gmail.com
```
should not be hardcoded either, use inputs.

dmitry-mightydevops replied:
combine two email steps into one `if: always()` and adjust your message based on job.status

kseniyashaydurova replied:
Don't see that it was fixed

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+on: [push]
+
+env:
+  REGISTRY: ghcr.io
```
pass it from inputs of the pipeline

kseniyashaydurova replied:
Don't see that it is fixed

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+
+
+      - name: Login to GitHub Container Registry
+        if: always()
```
you should not execute if tests failed
adjust in steps below as well

dmitry-mightydevops replied:
only always - is sending email.

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+
+      - name: Stop timer
+        if: always()
+        id: stop_timer
```
if you don't reference the id of the step in other step - don't introduce it, keep yaml file shorter.

dmitry-mightydevops commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+FROM node:18-alpine AS build
+WORKDIR /app
+COPY angular/package*.json ./
+RUN npm ci
+COPY . .
+WORKDIR /app/angular
+RUN npm run build
```
this is a duplicate unit of work. You build dist/ in the github actions - use it. Otherwise you double work yourself. One unit of work is in dockerfile (docker build) and second in the GHA itself. So use GHA and make simplest dockerfile based on nginx.

dmitry-mightydevops commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+WORKDIR /app/angular
+RUN npm run build
+
+FROM nginx:alpine
```
you version should always PINNED to SHA digest of the image to avoid building incompatible version in the future.

dmitry-mightydevops commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+FROM nginx:alpine
+COPY ./nginx.conf /etc/nginx/conf.d/default.conf
+COPY --from=build /app/angular/dist/angular-starter/browser /usr/share/nginx/html
+EXPOSE 80
```
good security principle is to always add USER command in the dockerfile to prevent running container under root user. 

It is possible that your base image doesn't have this (I'm sure it does) but some images don't (or they remove them in the future) - then your future build will allow root execution.

Also for compliance always include CMD command, when I open your dockerfile I need to know how you envision to the container to start, what command, I **may not** know the base image defaults and will not have time to go investigate base images.

dmitry-mightydevops replied:
always prefer port >1024
linux privileged ports are below 1024
so use 8080

On Linux, if a non-root user wants to run a service on a privileged port, they either need to use port forwarding (like using iptables) or elevate the service's privileges temporarily (e.g., via sudo or by using tools like setcap to grant specific capabilities to a binary)

dmitry-mightydevops commented on nginx.conf:
```diff
@@ -0,0 +1,9 @@
+        root /usr/share/nginx/html;
+        index index.html;
+        location / {
+            try_files $uri $uri/ /index.html =404;
```
this is unsafe solution
if your codebase contain sensitive files - you allow end users to render them, i.e. .env type of files.
check official docs what try_files does.

https://nginx.org/en/docs/http/ngx_http_core_module.html#try_files

find a better and safer way for your particular case.

kseniyashaydurova replied:
Don't see that it was fixed

dmitry-mightydevops reviewed (changes requested):
hi Denis, 

- pls adjust your step names to be lowercase
- pls add summary for your PRs (in real world it should contain what your changes are, how to use them, any context that is valuable for the reviewer)
- see my feedback below

dmitry-mightydevops commented:
also your pipeline runs failed
![image](https://github.com/user-attachments/assets/2a12cf92-16dd-4292-ad88-8d09979af3c7)

make sure to have one successful when you reopen the PR

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+  ci:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
```
Is it really needed? Can it work without it?

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+      packages: write
+
+    steps:
+      - name: Start timer
```
Pls separate each step with a new line for better readability

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+        if: always()
+        uses: docker/login-action@v1
+        with:
+          registry: ${{env.REGISTRY}}
```
Let's pass registry as input var

kseniyashaydurova replied:
Don't see that this comment is addressed

ErmakovDenis replied:
@kseniyashaydurova I looked at what the Input Var is represented, I did not understand what their convenience was.  Maybe there is some kind of example?

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+
+
+      - name: Build and push Docker image
+        if: always()
```
This also should be executed only if tests not failed

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+        run: |
+          echo "STOP_TIME=$(date +%s)" >> $GITHUB_ENV
+
+      - name: Check environment variables
```
Is this step needed? Can we omit it?

kseniyashaydurova replied:
What about this? Can you use env vars directly in body of the step where you send email?

kseniyashaydurova replied:
What about this? Can you use env vars directly in body of the step where you send email?

ErmakovDenis replied:
@kseniyashaydurova What does it mean?

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,103 @@
+          echo "EXECUTION_TIME_SECONDS=$SECONDS" >> $GITHUB_ENV
+          echo "Execution time: $HOURS hours $MINUTES minutes $SECONDS seconds"
+
+      - name: Send success email
```
Pls show us proofs with successful and failed emails screenshots

kseniyashaydurova reviewed (changes requested):
Cool, it really works! Let's improve codes a bit, added few more comments to @dmitry-mightydevops feedback.

ErmakovDenis committed:
Changed files according to comments

ErmakovDenis committed:
Try fix error get node.js version

ErmakovDenis committed:
Try fix error install dependencies

ErmakovDenis committed:
Try fix error install dependencies

ErmakovDenis committed:
Try fix error install dependencies

ErmakovDenis committed:
Try fix error install dependencies

ErmakovDenis committed:
Try fix error install dependencies

ErmakovDenis committed:
Trying without tests

ErmakovDenis committed:
Trying without tests 2

ErmakovDenis committed:
Trying without tests 3

ErmakovDenis committed:
Fix node version

ErmakovDenis committed:
Add nodejs version

ErmakovDenis committed:
Fix dockerfile

ErmakovDenis committed:
Fix dockerfile 2

ErmakovDenis committed:
Add build /dist in ci.yml

ErmakovDenis committed:
Screen success email

ErmakovDenis committed:
Final commit

ErmakovDenis commented:
Hi all! Thanks for the left comments, I tried to fix them all, but not everything worked out, let's understand:
I started by walking on all 'decorative' editors (corrected the names, removed the excess of the spaces)
Let's talk about Dockerfile, made it shorter, removed the first stage, now /Dist is going to CI, corrected the port. I completely agree with Dmitry regarding User, I checked that the container is not starting from under Root, nevertheless, I didn’t succeed in directly indicating it - I get mistakes, I hope you tell me what the reason may be.

I corrected nginx.conf - now it is safer, I hope I understood the comment correctly.

Probably the most offensive that ci.yml all the time falls into a mistake, the main reason is the tests, they fall by default. If we ignore this step, it would seem that everything should work, but not quite.
At the Build and Push Docker Image step, I get an error about the lack of rights to recording the container, and the reason is as follows, I use Github_token which works only in the Master branch, there is a bypass using the CR_PAT key, but it fundamentally does not want to work tremendously, despite the time that I get Success messages (I am attaching screenshots).
![image](https://github.com/user-attachments/assets/b119e22d-6f77-4828-80e4-52eef85d7249)
![image](https://github.com/user-attachments/assets/57c49cf5-fcaa-4574-bf50-9a8fa87d8844)

We’ll talk about hardcodding, in some places I consider it appropriate (in this work), so Dmitry wrote to take the Node.js version from the Package.json file, but in this repository the developer did not indicate it (do not need to be confused with Version, the nods versions are not recorded In such a format as here 18.2.1) (correctly 18.20.1)
For the examiner, I changed the value of the variable to 18.20.1 and pulled this value in CI.YML.

I also could not find how to use GH CLI, except that I only use additional extensions. (correct me).
Messages to the mail come correctly, I also attach screenshots.
![image](https://github.com/user-attachments/assets/f54209c0-145c-49d6-a28f-65970151b274)
![image](https://github.com/user-attachments/assets/3ad078ec-86c1-40cc-8fcf-11994a124b52)

I hope I have not forgotten anything.

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,112 @@
+          node-version: ${{steps.get-versions.outputs.node}}
+
+      - name: install dependencies
+        working-directory: ${{ github.workspace }}/angular
```
You have a lot of places where you repeat `working-directory` param. Let's specify it as default for the whole job

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,112 @@
+
+      - name: get node.js version from package.json
+        id: get-versions
+        run: echo "::set-output name=node::$(jq -r .version ${{ github.workspace }}/angular/package.json)"
```
LEt's use GITHUB_OUTPUT env var here? https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-an-output-parameter

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,112 @@
+
+      - name: install dependencies
+        working-directory: ${{ github.workspace }}/angular
+        run: npm install
```
It is better to use `npm ci` instead of `npm install`, cause it won't update packages versions - https://builtin.com/software-engineering-perspectives/npm-ci-vs-npm-install

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,112 @@
+        if: failure()
+        uses: dawidd6/action-send-mail@v3
+        with:
+          server_address: smtp.gmail.com
```
The same comments as above

kseniyashaydurova reviewed (changes requested):
Pls:
1. Add summary to your PR with task description
2. You should have 2 workflow run actions from your last commit before opening PR (1 - success, 2 - failed with error in tests run). Now I see only 1 failed CI/CD from your `Final commit`
![image](https://github.com/user-attachments/assets/01d52d0a-6d84-4c85-a94e-3a4a58bfcbbd)
3. Some feeback comments were not addressed, pls address them

ErmakovDenis committed:
Fix email notification and working directory

ErmakovDenis committed:
Fix working directory

ErmakovDenis committed:
Fix working directory 2

ErmakovDenis committed:
Fix working directory 3\

ErmakovDenis committed:
Fix working directory 4

ErmakovDenis committed:
Without tests

ErmakovDenis committed:
Test PAT

ErmakovDenis committed:
With tests

ErmakovDenis committed:
Fix docker compose (With tests)

ErmakovDenis committed:
Without tests

ErmakovDenis commented:
To begin with, @kseniyashaydurova asked me to make a successful commit without tests, and failed.
![image](https://github.com/user-attachments/assets/e5304425-059b-4ae2-b163-25cd433ec1c4)
Everything should be clearly visible.


It is also correct to work out Pat token, and I corrected the Docker compose file. (This is not so important)
It also seems to me that @kseniyashaydurova did not see the changes in the nginx file, correct me.

Corrected the latest comments of @kseniyashaydurova.
I still do not see how to apply the INPUT VAR and what is so bad virtual environment in this case.
I'm not sure that you see my comments

In the end, it would be for me to know what you generally think about work, it can be evaluated on a ten-point scale.
Thanks !

kseniyashaydurova reviewed (changes requested):
> To begin with, @kseniyashaydurova asked me to make a successful commit without tests, and failed. ![image](https://private-user-images.githubusercontent.com/108073309/376066917-e5304425-059b-4ae2-b163-25cd433ec1c4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg5NjMzMTYsIm5iZiI6MTcyODk2MzAxNiwicGF0aCI6Ii8xMDgwNzMzMDkvMzc2MDY2OTE3LWU1MzA0NDI1LTA1OWItNGFlMi1iMTYzLTI1Y2Q0MzNlYzFjNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxNVQwMzMwMTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZmU1ZWVlNDM0YWM1YzJiMGIzOTg5ODhiNjIwNmQ4MDViNjE0NDI0Y2VmZjAxM2NmMzRhNGNhZjljMjYwMGExJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Iq0UjYaHkSqa9bXZkYAclmuNGWiN1Yn8lUVn7_XxRBU) Everything should be clearly visible.
> 
> It is also correct to work out Pat token, and I corrected the Docker compose file. (This is not so important) It also seems to me that @kseniyashaydurova did not see the changes in the nginx file, correct me.
> 
> Corrected the latest comments of @kseniyashaydurova. I still do not see how to apply the INPUT VAR and what is so bad virtual environment in this case. I'm not sure that you see my comments
> 
> In the end, it would be for me to know what you generally think about work, it can be evaluated on a ten-point scale. Thanks !

1. I still don't see that all comments are addressed, pls, re request PR once you fix all comments
2. I saw your changes related to Nginx, the idea was to override the code to not use `try_files` directive and to find ways to make it work without it
3. We are asking to replace env vars to corresponding inputs, see docs - https://docs.github.com/en/enterprise-cloud@latest/actions/writing-workflows/workflow-syntax-for-github-actions#onworkflow_callinputs

Work can be evaluated once everything is fixed and looks nice :)

ErmakovDenis committed:
Add inputs vars

ErmakovDenis committed:
Check inputs

ErmakovDenis committed:
Check inpupts 2

ErmakovDenis committed:
Check inpupts 3

ErmakovDenis committed:
Check inpupts 4

ErmakovDenis committed:
Check inpupts 5

ErmakovDenis committed:
Check inpupts 6

ErmakovDenis committed:
Check inpupts 7

ErmakovDenis committed:
Try use inputs var

ErmakovDenis committed:
Try use inputs var 2

dmitry-mightydevops commented:
@ErmakovDenis I have converted PR to draft and add label wip
when all changes are addressed - pls re-request the PR and remove wip label and convert PR to "READY" status

ErmakovDenis committed:
test input vars

ErmakovDenis committed:
Add input vars

ErmakovDenis committed:
Delete test pipeline

ErmakovDenis committed:
Add input vars

ErmakovDenis committed:
Add input vars 2

ErmakovDenis committed:
Add input vars 3

ErmakovDenis committed:
Add new nginx, test workflow_dispatch

ErmakovDenis committed:
With tests

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,109 @@
+        run: |
+         echo "START_TIME=$(date +%s)" >> $GITHUB_ENV
+         echo The registry is ${{ inputs.registry }}
+         echo The server_address is ${{ inputs.server_address }}
```
let's simplifying the evaluation of time using
gh cli

there are methods to pull this information quickly 
gh accessible within the scripts of your pipeline
https://cli.github.com/

this should make this step unnecessary, as well as stop timer
and your calc time step would become very short.

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,109 @@
+
+      - uses: actions/setup-node@v1
+        with:
+          node-version: ${{steps.get-versions.outputs.node}}
```
let's instead use this method
![image](https://github.com/user-attachments/assets/4f5aee1d-bd68-475e-80ae-424937cc30b4)
https://github.com/actions/setup-node
https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#node-version-file

dmitry-mightydevops commented on Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM nginx@sha256:2140dad235c130ac861018a4e13a6bc8aea3a35f3a40e20c1b060d51a7efd250
+COPY ./nginx.conf /etc/nginx/conf.d/default.conf
+COPY ./angular/dist/angular-starter/browser /usr/share/nginx/html
+EXPOSE 8080
```
your dockerfile should
always have USER statement and CMD statement

dmitry-mightydevops commented on Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM nginx@sha256:2140dad235c130ac861018a4e13a6bc8aea3a35f3a40e20c1b060d51a7efd250
```
always put a link above referencing this revision by SHA1

dmitry-mightydevops commented on nginx.conf:
```diff
@@ -0,0 +1,15 @@
+
+  location ~* \.(js|css|png|jpg|jpeg|gif|svg|ico|woff|woff2|ttf|eot|otf)$ {
+      root /usr/share/nginx/html/;
+  }
```
just use "root" and "index" instructions for a single `/` location

ErmakovDenis committed:
Fix comments

ErmakovDenis committed:
Fix comments 1

ErmakovDenis committed:
Try gh

ErmakovDenis committed:
Try gh 1

ErmakovDenis committed:
Try gh 2

ErmakovDenis committed:
Try gh 2

kseniyashaydurova commented on Dockerfile:
```diff
@@ -0,0 +1,7 @@
+COPY ./angular/dist/angular-starter/browser /usr/share/nginx/html
+USER nginx
+CMD ["nginx", "-g", "daemon off;"]
+EXPOSE 8080
```
Usually it is good practise to finish dockerfile with CMD

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,111 @@
+  workflow_dispatch:
+    inputs:
+      registry:
+        description: "Registry"
```
Add better description everywhere, your description should always explain for what this param is needed

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,111 @@
+      registry:
+        description: "Registry"
+        required: false
+        default: "ghcr.io"
```
Remove `"`, in yaml files everything is a str and it just add noise

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,111 @@
+        type: string
+
+env:
+  GH_TOKEN: ${{ github.token }}
```
Do you really need to pass it as env? You can just use `{{ github.token }}` in the place where it is needed, without passing it as env var

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,111 @@
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        working-directory: ${{ github.workspace }}/angular
```
```suggestion
        working-directory: angular
```

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,111 @@
+
+      - uses: actions/setup-node@v4
+        with:
+          node-version-file: "angular/package.json"
```
No need in quotes too

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,111 @@
+        run: |
+          echo "START_TIME=$(date +%s)" >> $GITHUB_ENV
+
+      - uses: actions/setup-node@v4
```
Always add sensible names for all steps

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,111 @@
+      - name: build /dist
+        run: npm run build
+
+      #- name: run tests
```
Why no `tests` step?

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,111 @@
+        with:
+          registry: ${{ inputs.registry }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.PAT }}
```
You should use your env var instead

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,111 @@
+
+      - name: counting the time
+        if: always()
+        run: |
```
This is too complicated  implementation for such a simple logic, pls find out a way to make it one liner

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,111 @@
+          echo "EXECUTION_TIME_MINUTES=$MINUTES" >> $GITHUB_ENV
+          echo "EXECUTION_TIME_SECONDS=$SECONDS" >> $GITHUB_ENV
+
+      - name: counting the time gh cli
```
Why do you have that many steps for execution time calculation? Leave just one flow with GH CLI. Also don't see that results of this step are used

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,111 @@
+          body: |
+            Pipeline ${{ job.status }}!
+            Final execution time: ${{ env.EXECUTION_TIME_HOURS }} hours ${{ env.EXECUTION_TIME_MINUTES }} minutes ${{ env.EXECUTION_TIME_SECONDS }} seconds
+          to: clutcherdenis@gmail.com
```
Now you shared your email to the work and can get a SPAM

kseniyashaydurova commented on nginx.conf:
```diff
@@ -0,0 +1,15 @@
+  server_name localhost;
+
+  location / {
+    if (!-f $request_filename) {
```
Why that difficult? Can you use just `index` Nginx directive?

kseniyashaydurova reviewed (changes requested):
1. Still don't see task details in PR description.
2. Now application doesn't work
![image](https://github.com/user-attachments/assets/c9fffd63-b70c-4b17-a086-95b29ae9e9a0)

ErmakovDenis committed:
Try gh 3

ErmakovDenis committed:
Try gh 4

ErmakovDenis committed:
Fix comments

ErmakovDenis committed:
Fix comments

ErmakovDenis committed:
Add tests

ErmakovDenis commented:
Hi all!
At the moment, I have two problems:
1) I cannot forcibly create a user in a container
2) I can't make a timer in one line

I described the first point before (now nothing has changed), as for the second, you write to me to use gh cli, I studied the documentation, and did not find the right application.
The utility works great when the workflow is already completed, then I can conveniently and quickly get and use the data, but in my case I have to get the final time of the workflow, although it has not yet ended, maybe I don’t understand something, honestly I never thought that problems with the creation of a timer can arise :)

kseniyashaydurova commented:
> Hi all! At the moment, I have two problems:
> 
> 1. I cannot forcibly create a user in a container
> 2. I can't make a timer in one line
> 
> I described the first point before (now nothing has changed), as for the second, you write to me to use gh cli, I studied the documentation, and did not find the right application. The utility works great when the workflow is already completed, then I can conveniently and quickly get and use the data, but in my case I have to get the final time of the workflow, although it has not yet ended, maybe I don’t understand something, honestly I never thought that problems with the creation of a timer can arise :)

Hi Denis! Thanks for our feedback addressing, it is enough for test task. :) Don't worry about the rest comments!
